### PR TITLE
React Native Windows Support

### PR DESCRIPTION
**Pull request by ryanlntn**
Adds support for react-native-windows UWP and WPF.
Depends on upstream PR to RNW: https://github.com/ReactWindows/react-native-windows/pull/1050

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -95,7 +95,10 @@ react-native link react-native-view-shot
 
 #### Windows
 
-No support yet. Feel free to PR.
+1. In Visual Studio, in the solution explorer, right click on your project solution then select `Add` ➜ `ExisitingProject`
```
Comment (matthargett) [inline README.md]: s/project solution/solution

File: README.md [hunk 2]
```diff
@@ -95,7 +95,10 @@ react-native link react-native-view-shot
...
+2. Go to `node_modules` ➜ `react-native-view-shot` and add `RNViewShot.csproj` (UWP) or optionally `RNViewShot.Net46.csproj` (WPF)
+3. In Visual Studio, in the solution explorer, right click on your project then select `Add` ➜ `Reference`
```
Comment (matthargett) [inline README.md]: might want to say "Application project" here to be clear

File: windows/RNViewShot.Net46/RNViewShot.Net46.csproj [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{1DD45D00-DDF4-43B4-AB28-F98E6DC325B5}</ProjectGuid>
+    <OutputType>Library</OutputType>
+    <AppDesignerFolder>Properties</AppDesignerFolder>
+    <RootNamespace>RNViewShot.Net46</RootNamespace>
+    <AssemblyName>RNViewShot.Net46</AssemblyName>
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+    <FileAlignment>512</FileAlignment>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
```
Comment (matthargett) [inline windows/RNViewShot.Net46/RNViewShot.Net46.csproj]: need to get rid of the dreaded AnyCPU. x86, x64, and ARM only for RNW.

File: windows/RNViewShot/ViewShot.cs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                    }
+                }
+                else
+                {
+                    promise.Reject(ErrorUnableToSnapshot, "Unsupported result: " + result + ". Try one of: file | base64 | data-uri");
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.ToString());
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: use Debug.WriteLine() from the System.Diagnostics namespace instead.

File: windows/RNViewShot/ViewShot.cs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            }
+
+            await encoder.FlushAsync();
+
+            return encoder;            
+        }
+
+        private async Task<StorageFile> GetStorageFile()
+        {
+            StorageFolder storageFolder = ApplicationData.Current.LocalFolder;
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: var

File: windows/RNViewShot/ViewShot.cs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                            promise.Resolve(file.Path);
+                        }
+                    }
+                }
+                else if ("base64" == result)
+                {
+                    using (InMemoryRandomAccessStream ras = new InMemoryRandomAccessStream())
+                    {
+                        await CaptureView(view, ras);
+                        byte[] imageBytes = new byte[ras.Size];
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: var

File: windows/RNViewShot/ViewShot.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                        string data = Convert.ToBase64String(imageBytes);
+                        promise.Resolve(data);
+                    }
+                }
+                else if ("data-uri" == result)
+                {
+                    using (InMemoryRandomAccessStream ras = new InMemoryRandomAccessStream())
+                    {
+                        await CaptureView(view, ras);
+                        byte[] imageBytes = new byte[ras.Size];
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: var

File: windows/RNViewShot/RNViewShotModule.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            string path = options["path"] != null ? options.Value<string>("path") : null;
+
+            if (format != "png" && format != "jpg" && format != "jpeg")
+            {
+                promise.Reject(ViewShot.ErrorUnableToSnapshot, "Unsupported image format: " + format + ". Try one of: png | jpg | jpeg");
+                return;
+            }
+
+            UIManagerModule uiManager = this._reactContext.GetNativeModule<UIManagerModule>();
+            uiManager.AddUIBlock(new ViewShot(tag, format, quality, width, height, path, result, promise));
```
Comment (matthargett) [inline windows/RNViewShot/RNViewShotModule.cs]: extract new ViewShot to a variable for easier debugger inspection

File: windows/RNViewShot/ViewShot.cs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            }
+
+            try
+            {
+                if ("file" == result)
+                {
+                    using (InMemoryRandomAccessStream ras = new InMemoryRandomAccessStream())
+                    {
+                        await CaptureView(view, ras);
+                        StorageFile file = await GetStorageFile();
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: var

File: windows/RNViewShot/ViewShot.cs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+                        await CaptureView(view, ras);
+                        byte[] imageBytes = new byte[ras.Size];
+                        await ras.AsStream().ReadAsync(imageBytes, 0, imageBytes.Length);
+                        string data = Convert.ToBase64String(imageBytes);
+                        promise.Resolve(data);
+                    }
+                }
+                else if ("data-uri" == result)
+                {
+                    using (InMemoryRandomAccessStream ras = new InMemoryRandomAccessStream())
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: var

File: windows/RNViewShot/ViewShot.cs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        private async Task<StorageFile> GetStorageFile()
+        {
+            StorageFolder storageFolder = ApplicationData.Current.LocalFolder;
+            if (string.IsNullOrEmpty(path))
+            {
+                string fileName = Guid.NewGuid().ToString();
+                fileName = Path.ChangeExtension(fileName, extension);
+                return await storageFolder.CreateFileAsync(fileName, CreationCollisionOption.ReplaceExisting);
```
Comment (matthargett) [inline windows/RNViewShot/ViewShot.cs]: why not assign path here and then ditch the else{} wrapper on the next line?

**Comment by gre** [comment]
incredible PR. How awesome is that!
Also thanks for reviewing it @matthargett .

**Comment by gre** [comment]
@ryanlntn is the PR ready to be merged now? :)

**Comment by ryanlntn** [comment]
@gre Yes but you might want to hold off until https://github.com/ReactWindows/react-native-windows/pull/1050 is merged since it won't work without that.

**Comment by gre** [comment]
@ryanlntn ok nice!

BTW if you want to have more testing, maybe it would be great to make that https://github.com/gre/react-native-view-shot-example a windows app too. it's basically an app I use to test the library and it should cover all different options / different kind of views.

**Comment by ryanlntn** [comment]
@gre Planning on it! 😄

**Comment by gre** [comment]
It's funny you had to do the same third party feature for the same purpose hehe

**Comment by ryanlntn** [comment]
@gre Yeah a blatant copy actually. 😛

**Comment by ryanlntn** [comment]
@gre ReactWindows/react-native-windows#1050 has landed in RNW master.

**Comment by gre** [comment]
so cool! I'll probably get this merged anyway :) thanks again!

**Pull request merged**